File: backend-spring-boot/src/test/java/no/vicx/backend/jwt/github/vm/GitHubUserResponseVmTest.java
package no.vicx.backend.jwt.github.vm;

import org.junit.jupiter.api.Test;
import org.springframework.security.core.authority.SimpleGrantedAuthority;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.contains;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;

class GitHubUserResponseVmTest {

    @Test
    void toPrincipal_givenUserWithAllFieldsSet_expectPrincipalWithAllFieldsSet() {
        var sut = new GitHubUserResponseVm(
                USER_IN_TEST, "~granted-scopes~", "~token~");

        var principal = sut.toPrincipal();

        assertNotNull(principal);
        assertEquals("~login~", principal.getName());
        assertThat(
                principal.getAuthorities(),
                contains(new SimpleGrantedAuthority("ROLE_GITHUB_USER")));
    }

    private static final GitHubUserVm USER_IN_TEST = new GitHubUserVm(
            "~id~", "~login~", "~name~", "~email~", "~avatar~");
}

File: backend-spring-boot/src/test/kotlin/no/vicx/backend/jwt/github/vm/GitHubUserResponseVmTest.kt
package no.vicx.backend.jwt.github.vm

import io.kotest.assertions.assertSoftly
import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.matchers.collections.shouldContain
import io.kotest.matchers.shouldBe
import no.vicx.backend.jwt.github.GitHubTestUtils.githubUserInTest
import org.springframework.security.core.authority.SimpleGrantedAuthority

class GitHubUserResponseVmTest : BehaviorSpec({

    Given("a GitHubUserResponseVm with all fields set") {
        val sut = GitHubUserResponseVm(
            user = githubUserInTest,
            grantedScopes = "~granted-scopes~",
            token = "~token~"
        )

        When("calling toPrincipal") {
            val principal = sut.toPrincipal()

            Then("principal contains correct name and roles") {
                assertSoftly(principal) {
                    name shouldBe githubUserInTest.login
                    authorities shouldContain SimpleGrantedAuthority("ROLE_GITHUB_USER")
                }
            }
        }
    }
})